qa/tasks/cephfs/test_exports.py

```diff
@@ -198,69 +198,68 @@ def test_ephemeral_pin_dist_override(self):
 
     def test_ephemeral_pin_dist_override_pin(self):
         """
         That an export pin overrides an ephemerally pinned directory.
         """
 
-        self._setup_tree(distributed=True, export=0)
-        subtrees = self._wait_distributed_subtrees(100, status=self.status, rank="all", path="/tree/")
-        which = None
-        for s in subtrees:
-            if s['auth_first'] == 1:
-                path = s['dir']['path']
-                self.mount_a.setfattr(path[1:], "ceph.dir.pin", "0")
-                which = path
-                break
-        self.assertIsNotNone(which)
+        self._setup_tree(distributed=True)
+        subtrees = self._wait_distributed_subtrees(3 * 2, status=self.status, rank="all")
+        self.mount_a.setfattr("tree", "ceph.dir.pin", "0")
         time.sleep(15)
         subtrees = self._get_subtrees(status=self.status, rank=0)
         for s in subtrees:
             path = s['dir']['path']
-            if path == which:
+            if path == '/tree':
                 self.assertEqual(s['auth_first'], 0)
                 self.assertFalse(s['distributed_ephemeral_pin'])
-                return
         # it has been merged into /tree
 
     def test_ephemeral_pin_dist_off(self):
         """
         That turning off ephemeral distributed pin merges subtrees.
         """
 
-        self._setup_tree(distributed=True, export=0)
-        self._wait_distributed_subtrees(100, status=self.status, rank="all")
+        self._setup_tree(distributed=True)
+        self._wait_distributed_subtrees(3 * 2, status=self.status, rank="all")
         self.mount_a.setfattr("tree", "ceph.dir.pin.distributed", "0")
-        self._wait_subtrees([('/tree', 0)], status=self.status)
+        time.sleep(15)
+        subtrees = self._get_subtrees(status=self.status, rank=0)
+        for s in subtrees:
+            path = s['dir']['path']
+            if path == '/tree':
+                self.assertFalse(s['distributed_ephemeral_pin'])
+
 
     def test_ephemeral_pin_dist_conf_off(self):
         """
         That turning off ephemeral distributed pin config prevents distribution.
         """
 
-        self._setup_tree(export=0)
+        self._setup_tree()
         self.config_set('mds', 'mds_export_ephemeral_distributed', False)
         self.mount_a.setfattr("tree", "ceph.dir.pin.distributed", "1")
-        time.sleep(30)
-        self._wait_subtrees([('/tree', 0)], status=self.status)
+        time.sleep(15)
+        subtrees = self._get_subtrees(status=self.status, rank=0)
+        for s in subtrees:
+            path = s['dir']['path']
+            if path == '/tree':
+                self.assertFalse(s['distributed_ephemeral_pin'])
 
-    def test_ephemeral_pin_dist_conf_off_merge(self):
+    def _test_ephemeral_pin_dist_conf_off_merge(self):
         """
         That turning off ephemeral distributed pin config merges subtrees.
+        FIXME: who triggers the merge?
```

Comment:
?

Reply:
the test relies on bug in CInode::maybe_export_pin.